### PR TITLE
WIP compile ahead of time

### PR DESCRIPTION
**Pull request by rbavery**
This PR seeks to resolve the long compile step documented in https://github.com/aquacropos/aquacrop/issues/17

Primary goal is to replace JIT compilation with AOT compilation, still using numba: https://numba.pydata.org/numba-doc/dev/user/pycc.html

It also adds test for run time of the Tunis wheat simulation example and the time to import the module (so far used to produce these timings https://github.com/aquacropos/aquacrop/issues/17#issuecomment-1019607858)

### CODE REVIEW

**Comment by thomasdkelly** [comment]
commented out the jit decorators
added cc.export to 2 functions (growing_degree_days, drainage) and compiled locally
the new compiled module is also included in this commit
slight changes to the speed tests

The workflow i used was:

1. Make changes to function arguments and variables
2. add cc.export decorator
3. remove function name from `__all__ ` at top of classes.py
4. run solution.py to compile
5. go to timestep.py and import the new function explicity from the new compiled module soution_aot
6. change any function arguments that need changing
7. run test functions

**Comment by thomasdkelly** [comment]
Have also added the tests to the github aciton so we can keep can also now keep track how much faster the sims are getting as we process more functions

**Comment by thomasdkelly** [comment]
Since there are quite a few functions to change inside solution.py i figured it would be create a checlist so that we dont try and change the same ones

## Change commited

- [ ] growing_degree_day
- [ ] drainage

## Have started changing locally

## To do

- [ ] root_zone_water
- [ ] check_groundwater_table
- [ ] root_development
- [ ] pre_irrigation
- [ ] rainfall_partition
- [ ] irrigation
- [ ] infiltration
- [ ] capillary_rise
- [ ] germination
- [ ] growth_stage
- [ ] water_stress
- [ ]  cc_development
- [ ] cc_required_time
- [ ] adjust_CCx
- [ ]  update_CCx_CDC
- [ ] canopy_cover
- [ ]  evap_layer_water_content
- [ ] soil_evaporation
- [ ] aeration_stress
- [ ] transpiration
- [ ] groundwater_inflow
- [ ] HIref_current_day
- [ ] biomass_accumulation
- [ ] temperature_stress
- [ ] HIadj_pre_anthesis
- [ ] HIadj_pollination
- [ ] HIadj_post_anthesis
- [ ] harvest_index

**Comment by thomasdkelly** [comment]
compiling root_zone_water had a big impact as it is called many times each timestep. Import and simulation times are starting to look good

**Comment by rbavery** [comment]
@thomasdkelly I can't edit the comment or add check marks but I'm starting from the top of the todo list

**Comment by thomasdkelly** [comment]
Just done a couple more, soil_evaporation was a long one but all works. 

Also used PyCallGraph to visualize how long each function is taking to run. Was quite interesting to see. 

Also found that if the function we want to cc.export includes another cc.export function within it, the inner function needs the @njit decorator above it.

**Comment by thomasdkelly** [comment]
As a point of reference here is the test time outputs i get on my pc for the current jit master branch:

compile time: 26.046
total sim time for 1 repetitions: 1.417
total sim time for 10 repetitions: 1.709

Compared to the latest commit to this branch with AOT:

compile time: 0.0
total sim time for 1 repetitions: 0.337
total sim time for 10 repetitions: 3.36

The compile time would be longer ~0.3s but i removed the `__init__` compile function as it is no longer needed with AOT.

So hopefully we can get close to or exceed that 1.7s number with AOT, then we know we are doing okay

**Comment by thomasdkelly** [comment]
So after seeing the following discussions and docs, structrefs and namedtuples were suggested:

- https://numba.discourse.group/t/whats-the-recommended-way-to-pass-around-large-amounts-of-parameters-that-works-with-aot/1154
- https://numba.discourse.group/t/any-customizable-container-that-doesnt-box-unbox-when-passed-around/340
- https://numba.pydata.org/numba-doc/dev/extending/high-level.html#implementing-mutable-structures


The only data structure within aquacrop that needs to be mutable is NewCond as this tracks every relavent and changing variables as the simulation runs. Everything else (Crop, Soil.Profile, FieldMngt, IrrMngt,etc.) could be instead be defined as a namedtuple. The code is a lot cleaner and shorter for namedtuple than for structref. so may be a good idea. I have tested that the example function can be AOT compiled and run as expected:
```
@cc.export('change_with_named_tup',(mystruct_type_sig,mynamedtup_type_sig))
def change_with_named_tup(mystructref,mynamedtup):

    mystructref.val = mynamedtup.val2*7
    
    return mystructref

cc.compile()

```
So will create a seperate branch and pull request building off of this one (just incase other issues arrise and we have to go back to passing just floats and arrays again)


What do you think @rbavery

**Comment by rbavery** [comment]
From our meeting, just recapping that we'll be using structrefs since they are mutable and support both JIT and AOT compilation. This will allow us to move to AOT, and be able to move back to JIt for some or all parts of the code if there's a reason to (though we don't expect JIt-specific optimizations to matter)

**Comment by arongergely** [comment]
Hi I would like to help move the rest of the functions to AOT. Are you still carrying on with the same idea?

If I may jump in, moving any *invariables* to immutable objects seems logical. AFAIK tuples are still faster to access and have less footprint in python than any mutables

**Comment by thomasdkelly** [comment]
> From our meeting, just recapping that we'll be using structrefs since they are mutable and support both JIT and AOT compilation. This will allow us to move to AOT, and be able to move back to JIt for some or all parts of the code if there's a reason to (though we don't expect JIt-specific optimizations to matter)

Yes both structref and namedtuples can be jit and AOT compiled. For the reasons that @arongergely has given as well as the significant amount of code required to define the structref taken from my links above:

```
# Define a StructRef.
# `structref.register` associates the type with the default data model.
# This will also install getters and setters to the fields of
# the StructRef.
@structref.register
class MyStructType(types.StructRef):
    def preprocess_fields(self, fields):
        # This method is called by the type constructor for additional
        # preprocessing on the fields.
        # Here, we don't want the struct to take Literal types.
        return tuple((name, types.unliteral(typ)) for name, typ in fields)


# Define a Python type that can be use as a proxy to the StructRef
# allocated inside Numba. Users can construct the StructRef via
# the constructor for this type in python code and jit-code.
class MyStruct(structref.StructRefProxy):
    def __new__(cls, name, vector):
        # Overriding the __new__ method is optional, doing so
        # allows Python code to use keyword arguments,
        # or add other customized behavior.
        # The default __new__ takes `*args`.
        # IMPORTANT: Users should not override __init__.
        return structref.StructRefProxy.__new__(cls, name, vector)

    # By default, the proxy type does not reflect the attributes or
    # methods to the Python side. It is up to users to define
    # these. (This may be automated in the future.)

    @property
    def name(self):
        # To access a field, we can define a function that simply
        # return the field in jit-code.
        # The definition of MyStruct_get_name is shown later.
        return MyStruct_get_name(self)

    @property
    def vector(self):
        # The definition of MyStruct_get_vector is shown later.
        return MyStruct_get_vector(self)


@njit
def MyStruct_get_name(self):
    # In jit-code, the StructRef's attribute is exposed via
    # structref.register
    return self.name


@njit
def MyStruct_get_vector(self):
    return self.vector


# This associates the proxy with MyStructType for the given set of
# fields. Notice how we are not contraining the type of each field.
# Field types remain generic.
structref.define_proxy(MyStruct, MyStructType, ["name", "vector"])

``` 
We would have to define two functions for every attribute of the structref. This will of course just be lots of copy-paste and only has to be done once. However it is quite long and confusing at first look. In comparison, all that is needed to define the named tuple is:
```
nt2 = namedtuple('nt2', 'x,y,z')
nt2_typ_sig= types.NamedTuple((types.int64,types.float64,types.float64[:]),nt2)

```

So using namedtuples for everything except InitCondClass and IrrMngtClass I think is the best idea. If we ever want the ability to mutate other variables later down the line (e.g. FieldMngt) we can simply change from namedtuple to struct, or find a way to just use the already mutable InitCondClass. 

I will get the ball rolling today by defining a namedtuple for `Soil.Profile` (which gets passed around alot). And a structref for `NewCond` (InitCondClass). From there we can carry on working through all the functions in solution.py. 

@arongergely Thanks for your offer of help. Until I define the structrefs and namedtuples (should be by the end of today). Perhaps you could do some of the functions that already just take in floats/arrays/strings.  For example a lot of the canopy cover functions are like this (update_CCx_CDC, ...)

**Comment by arongergely** [comment]
> @arongergely Thanks for your offer of help. Until I define the structrefs and namedtuples (should be by the end of today), perhaps you could do some of the functions that already just take in floats/arrays/strings. For example a lot of the canopy cover functions are like this (update_CCx_CDC, ...)

Thank you. I am taking these: `cc_development`, `cc_required_time`, `adjust_CCx`, `update_CCx_CDC`

**Comment by thomasdkelly** [comment]
Have been having a pretty rough time with this the past 2 days. Here is a list of problems encountered

- Defining the structref for the NewCond class takes an insanse amount of lines of code. For every attribute (of which there are alot!) you must create two seperate functions as stated above:
```

@property
def vector(self):
    # The definition of MyStruct_get_vector is shown later.
    return MyStruct_get_vector(self)

@njit
def MyStruct_get_vector(self):
    return self.vector

```

- Next is that after all this you cannot mutate the attributes within python code. You can only do it inside numba compiled code. This creates a whole host of problems as there are many parts of the code that we are not interested/able to compile that need to access and mutate the NewCond structure.
- After these issues I instead tried to focus on getting all the other functions aot compiled and converting classess into namedtuples. This works okay for some (CropClass) but had some issues I couldnt resolve when creating the `thRZ` named tuple inside the aot ccompiled `root_zone_water` function. Was getting a segmentation error.
- Moving beyond that I just wanted to compile some more functions and ignore the namedtuples (i.e. just take in floats/arrays as arguments) and so tried to compile the `irrigation` function. This function calls the already compiled `root_zone_water` function. So was trying to compile a function that contains a call to another compiled function.  This does run okay however requires the inner function to also have the @njit decorator (https://github.com/numba/numba/issues/5910). This means that we now have a large (1-2 second) compilaiton time. So we are actually getting faster run times by not compiling the `irrigation` function.
- So this issue will also be the same for functions like `transpiration` and `harvest_index` that similarly want to call other compiled functions within them. This is something we potentially have to accept for now (and just see what kind of speeds we can achieve). I have noticed that some of these inner function calls do not have to take place within the function itself as they do not rely on any newly calculated variables. So perhaps ill have to take a closer look at some of these.


So to sum up: Incorperating AOT compilation + Structref + namedtuple is not as straightforward as previously hoped. I propose for this pull request we continue AOT compiling any functions that do not call any inner functions, Ignore structrefs for now and namedtuples (except the namedtuples for `FieldMngt` and `Crop`). This means passing lots of `NewCond_something` arguments but this is okay for now. 

Once done we will have something that has zero import/compilation time,  >5x quicker for single simulations, and maybe 1.5x slower over many simulations. 

That should be a good point to merge and create the first release. And then carry on further speed improvements from there.

What do you guys think

**Comment by arongergely** [comment]
Thanks for the status and thoughts @thomasdkelly. I also ran into the  compile issues you mentioned when trying to move `adjust_CCx`, which would call functions that I have just AOT compiled. It's still giving me lemons, the **@njit** decorator does not solve it for me. 

Perhaps it's an idea then to leave out the ones which call AOT-compiled functions out this for now. 
Or otherwise we could iron out the structure bit as you said, so functions within `solutions.py` would not call each other, therefore could be independently compiled. 

I have  moved to AOT the other 3 I took: `cc_development`, `cc_required_time`, `update_CCx_CDC`.

How to get my changes added in here? Should issue a PR to this branch on @rbavery 's  fork?

**Comment by rbavery** [comment]
> I propose for this pull request we continue AOT compiling any functions that do not call any inner functions

I'm on board with your proposal @thomasdkelly thanks for digging into this and documenting the issues you faced!

**Comment by thomasdkelly** [comment]
Thanks both for your help, will merge and create some additional issues/prs to make further speed improvements. Next priority is to create a github release (v0.3) and create a github action to upload releases automatically to pypi
